{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "6a5709f9-8dd8-4d51-84a5-64398588df2e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Import the pandas library for data and analysis manipulation.\n",
    "# Assign pandas library to a shorter alias 'pd'.\n",
    "\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "b4823962-4ca6-4b28-be09-cf1d0377b911",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Define an empty dictionary named 'dtype_dict'\n",
    "# This dictionary will map certain data attributes to their respective data types.\n",
    "\n",
    "dtype_dict = {\n",
    "    \"vcf_filter\": str, # Data in \"vcf_filter\" column should be treated as type 'string' (text).\n",
    "    \"vcf_mutect\": float, # Data in \"vcf_count_mutect\" should be treated as floating-point numbers.\n",
    "    \"cosmic_haem\": str,  # Data in \"cosmic_haem\" column should be treated as type 'string' (text).\n",
    "    \"vcf_pindel\": float, # Data in \"vcf_count_pindel\" should be treated as floating-point numbers.\n",
    "    \"vaf\": float, # Data in \"vaf\" should be treated as floating-point numbers.\n",
    "    \"depth\": float, # Data in \"depth\" should be treated as floating-point numbers.\n",
    "    \"gnomAD_AF\": float, # Data in \"gnomad_AF\" should be treated as floating-point numbers. \n",
    "    \"genes\": str, # Data in \"gene_name\" column should be treated as type 'string' (text).\n",
    "    \"variants\": str, # Data in \"variant\" column should be treated as type 'string' (text).\n",
    "    \"var_type\": str, # Data in \"var_type\" column should be treated as type 'string' (text).\n",
    "    \"virtual_panel\": str, # Data in \"virtual_panel\" column should be treated as type 'string' (text).\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "49447a3a-eb07-44f8-aecd-3759bba4d750",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load a CSV file into a pandas DataFrame.\n",
    "# Specify additional parameters and store the resulting DataFrame in the variable 'data'.\n",
    "\n",
    "# The 'read_csv' function is used to load the file 'mock_validation_rules.csv'.\n",
    "data = pd.read_csv(r'C:\\Users\\PC\\Desktop\\NGS-Mock-Variant-Analysis\\4.data\\mock_validation_rules.csv', \n",
    "                   # This disables low memory mode and allows pandas to read the entire file into memory, even if it's very large.\n",
    "                   low_memory=False,  \n",
    "                   # 'dtype_dict' specifies the data types for specific columns in the CSV file.\n",
    "                   dtype=dtype_dict) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "47721990-c3f8-4c67-ba5f-366ab471b50a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_id</th>\n",
       "      <th>variants</th>\n",
       "      <th>genes</th>\n",
       "      <th>cNomen</th>\n",
       "      <th>pNomen</th>\n",
       "      <th>run</th>\n",
       "      <th>vaf</th>\n",
       "      <th>depth</th>\n",
       "      <th>vcf_mutect</th>\n",
       "      <th>vcf_pindel</th>\n",
       "      <th>...</th>\n",
       "      <th>gnomAD_AF</th>\n",
       "      <th>gnomAD_AF_Pct</th>\n",
       "      <th>cosmic_id</th>\n",
       "      <th>cosmic_haem</th>\n",
       "      <th>run_freq</th>\n",
       "      <th>total_freq</th>\n",
       "      <th>length</th>\n",
       "      <th>var_type</th>\n",
       "      <th>technical_status</th>\n",
       "      <th>retain_discard</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Mock001</td>\n",
       "      <td>KRAS_2_160883814_A/G</td>\n",
       "      <td>KRAS</td>\n",
       "      <td>c.1077A&gt;G</td>\n",
       "      <td>p.Ala1367Val</td>\n",
       "      <td>Run1</td>\n",
       "      <td>0.377</td>\n",
       "      <td>354.0</td>\n",
       "      <td>22.0</td>\n",
       "      <td>26.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.019014</td>\n",
       "      <td>1.9014</td>\n",
       "      <td>COSM444461</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2</td>\n",
       "      <td>86</td>\n",
       "      <td>0</td>\n",
       "      <td>deletion</td>\n",
       "      <td>Discard</td>\n",
       "      <td>Discard</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Mock002</td>\n",
       "      <td>KRAS_20_183670202_A/G</td>\n",
       "      <td>KRAS</td>\n",
       "      <td>c.2772A&gt;G</td>\n",
       "      <td>p.Ala228Val</td>\n",
       "      <td>Run1</td>\n",
       "      <td>0.727</td>\n",
       "      <td>254.0</td>\n",
       "      <td>29.0</td>\n",
       "      <td>27.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.011973</td>\n",
       "      <td>1.1973</td>\n",
       "      <td>COSM989868</td>\n",
       "      <td>6.0</td>\n",
       "      <td>3</td>\n",
       "      <td>84</td>\n",
       "      <td>0</td>\n",
       "      <td>indel</td>\n",
       "      <td>Valid</td>\n",
       "      <td>Retain</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Mock003</td>\n",
       "      <td>KRAS_22_5738866_A/G</td>\n",
       "      <td>KRAS</td>\n",
       "      <td>c.2073A&gt;G</td>\n",
       "      <td>p.Ala218Val</td>\n",
       "      <td>Run3</td>\n",
       "      <td>0.163</td>\n",
       "      <td>291.0</td>\n",
       "      <td>41.0</td>\n",
       "      <td>35.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.003120</td>\n",
       "      <td>0.3120</td>\n",
       "      <td>COSM924050</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0</td>\n",
       "      <td>80</td>\n",
       "      <td>0</td>\n",
       "      <td>deletion</td>\n",
       "      <td>Discard</td>\n",
       "      <td>Discard</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Mock004</td>\n",
       "      <td>BRCA2_18_377694477_A/G</td>\n",
       "      <td>BRCA2</td>\n",
       "      <td>c.2030A&gt;G</td>\n",
       "      <td>p.Ala947Val</td>\n",
       "      <td>Run3</td>\n",
       "      <td>0.067</td>\n",
       "      <td>619.0</td>\n",
       "      <td>39.0</td>\n",
       "      <td>50.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.017324</td>\n",
       "      <td>1.7324</td>\n",
       "      <td>COSM606687</td>\n",
       "      <td>2.0</td>\n",
       "      <td>3</td>\n",
       "      <td>68</td>\n",
       "      <td>0</td>\n",
       "      <td>substitution</td>\n",
       "      <td>Valid</td>\n",
       "      <td>Discard</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Mock005</td>\n",
       "      <td>ARID1A_15_547373809_A/G</td>\n",
       "      <td>ARID1A</td>\n",
       "      <td>c.382A&gt;G</td>\n",
       "      <td>p.Ala782Val</td>\n",
       "      <td>Run2</td>\n",
       "      <td>0.599</td>\n",
       "      <td>718.0</td>\n",
       "      <td>33.0</td>\n",
       "      <td>14.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.014161</td>\n",
       "      <td>1.4161</td>\n",
       "      <td>COSM938076</td>\n",
       "      <td>9.0</td>\n",
       "      <td>0</td>\n",
       "      <td>11</td>\n",
       "      <td>0</td>\n",
       "      <td>indel</td>\n",
       "      <td>Discard</td>\n",
       "      <td>Retain</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 24 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  sample_id                 variants   genes     cNomen        pNomen   run  \\\n",
       "0   Mock001     KRAS_2_160883814_A/G    KRAS  c.1077A>G  p.Ala1367Val  Run1   \n",
       "1   Mock002    KRAS_20_183670202_A/G    KRAS  c.2772A>G   p.Ala228Val  Run1   \n",
       "2   Mock003      KRAS_22_5738866_A/G    KRAS  c.2073A>G   p.Ala218Val  Run3   \n",
       "3   Mock004   BRCA2_18_377694477_A/G   BRCA2  c.2030A>G   p.Ala947Val  Run3   \n",
       "4   Mock005  ARID1A_15_547373809_A/G  ARID1A   c.382A>G   p.Ala782Val  Run2   \n",
       "\n",
       "     vaf  depth  vcf_mutect  vcf_pindel  ... gnomAD_AF gnomAD_AF_Pct  \\\n",
       "0  0.377  354.0        22.0        26.0  ...  0.019014        1.9014   \n",
       "1  0.727  254.0        29.0        27.0  ...  0.011973        1.1973   \n",
       "2  0.163  291.0        41.0        35.0  ...  0.003120        0.3120   \n",
       "3  0.067  619.0        39.0        50.0  ...  0.017324        1.7324   \n",
       "4  0.599  718.0        33.0        14.0  ...  0.014161        1.4161   \n",
       "\n",
       "    cosmic_id cosmic_haem  run_freq  total_freq length      var_type  \\\n",
       "0  COSM444461         0.0         2          86      0      deletion   \n",
       "1  COSM989868         6.0         3          84      0         indel   \n",
       "2  COSM924050         2.0         0          80      0      deletion   \n",
       "3  COSM606687         2.0         3          68      0  substitution   \n",
       "4  COSM938076         9.0         0          11      0         indel   \n",
       "\n",
       "   technical_status  retain_discard  \n",
       "0           Discard         Discard  \n",
       "1             Valid          Retain  \n",
       "2           Discard         Discard  \n",
       "3             Valid         Discard  \n",
       "4           Discard          Retain  \n",
       "\n",
       "[5 rows x 24 columns]"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Display the first few rows of the DataFrame using the 'head()' method without argument to show first 5 rows by default.\n",
    "# This is neccessary to quickly check if the data was loaded correctly.\n",
    "\n",
    "data.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "caeba265-c709-4e0f-b2d8-8bb49da5c123",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "100\n"
     ]
    }
   ],
   "source": [
    "# Check the number of rows in the DataFrame.\n",
    "\n",
    "print(len(data))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "3022a4e9-e4bf-4d14-8905-503b20310824",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Lenght of rows with 'valid' technical_status but incorrect retain_discard:\n",
      "0\n"
     ]
    }
   ],
   "source": [
    "# Find rows where technical_status does not match the 'retain_discard' column.\n",
    "# 1. Filter the rows where technical_status is 'valid' and retain_discard is **not** 'Retain'.\n",
    "# Combine the conditions and store the filtered rows in a new DataFrame named `valid_mismatch`.\n",
    "\n",
    "valid_mismatch = data[(data['technical_status'] == 'valid') & (data['retain_discard'] != 'Retain')]\n",
    "\n",
    "# Calculate the number of rows in `valid_mismatch`.\n",
    "\n",
    "print(\"Lenght of rows with 'valid' technical_status but incorrect retain_discard:\")\n",
    "print(len(valid_mismatch))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "6fb2be9e-efc5-46be-9645-be769af0f619",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Lenght of rows with 'discarded' technical_status but incorrect retain_discard:\n",
      "0\n"
     ]
    }
   ],
   "source": [
    "# 2. Filter the rows where technical_status is 'discarded' and retain_discard is **not** 'Retain' or 'Unknown'.\n",
    "# The `.isin([...])` check `retain_discard` is in the list and the `~` select rows where `retain_discard` is **not** in `['Discard', 'Unknown']`.\n",
    "# Combine the conditions and store the filtered rows in a new DataFrame named `discarded_mismatch`.\n",
    "\n",
    "discarded_mismatch = data[(data['technical_status'] == 'discarded') & (~data['retain_discard'].isin(['Discard', 'Unknown']))]\n",
    "\n",
    "# Calculate the number of rows in `discarded_mismatch`.\n",
    "\n",
    "print(\"Lenght of rows with 'discarded' technical_status but incorrect retain_discard:\")\n",
    "print(len(discarded_mismatch))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "b05a96cb-615c-44ab-8329-c6e0e79ca3be",
   "metadata": {},
   "outputs": [],
   "source": [
    "# 1. Define a function named `status_match` that takes a `row` as input.\n",
    "\n",
    "   #   ii. If `technical_status` is `'discarded'`, then `retain_discard` must be in the list `['Discard', 'Unknown']`.\n",
    "\n",
    "# Define equivalence rules for comparison\n",
    "def status_match(row):\n",
    "    # Check the conditions for equivalence:\n",
    "    # i. If `technical_status` is `'valid'`, then `retain_discard` must be `'Retain'`.\n",
    "    if row['technical_status'] == 'valid' and row['retain_discard'] == 'Retain':\n",
    "        # Return `True` if the equivalence rule is satisfied.\n",
    "        return True\n",
    "    # ii. Also check if `technical_status` is `'discarded'`, then `retain_discard` must be in the list `['Discard', 'Unknown']`.\n",
    "    elif row['technical_status'] == 'discarded' and row['retain_discard'] in ['Discard', 'Unknown']:\n",
    "        # Return `True` if the equivalence rule is satisfied.\n",
    "        return True\n",
    "    # Return `False` if the equivalence rule is not met.\n",
    "    return False"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "c4b2d323-ef4f-42df-9540-f92a803f0652",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Store the result in a new column 'status_match' in the DataFrame.\n",
    "# Use the 'apply' method to apply the matching rule 'status_match' to each row.\n",
    "# The 'axis=1' argument specifies that the function should be applied row-wise, as (axis=0 would apply it to columns).\n",
    "\n",
    "data['status_match'] = data.apply(status_match, axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "0b4696e0-d914-48c1-9982-32fc52bc6dfe",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Filter all the rows where the statuses match.\n",
    "# Select rows where `status_match` is `True` and store these rows in a new DataFrame named `mismatch_rows`.\n",
    "\n",
    "match_rows = data[data['status_match'] == True]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "d0943089-1a50-4d94-a48c-c12d9d3d8719",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Lenght of all the rows where the statuses match:\n",
      "0\n"
     ]
    }
   ],
   "source": [
    "# Check the number of all the rows where the statuses match in the DataFrame.\n",
    "\n",
    "print(\"Lenght of all the rows where the statuses match:\")\n",
    "print(len(match_rows))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "0dae92f5-dd74-43ff-9abe-9eaf5b53d4f3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_id</th>\n",
       "      <th>variants</th>\n",
       "      <th>genes</th>\n",
       "      <th>cNomen</th>\n",
       "      <th>pNomen</th>\n",
       "      <th>run</th>\n",
       "      <th>vaf</th>\n",
       "      <th>depth</th>\n",
       "      <th>vcf_mutect</th>\n",
       "      <th>vcf_pindel</th>\n",
       "      <th>...</th>\n",
       "      <th>gnomAD_AF_Pct</th>\n",
       "      <th>cosmic_id</th>\n",
       "      <th>cosmic_haem</th>\n",
       "      <th>run_freq</th>\n",
       "      <th>total_freq</th>\n",
       "      <th>length</th>\n",
       "      <th>var_type</th>\n",
       "      <th>technical_status</th>\n",
       "      <th>retain_discard</th>\n",
       "      <th>status_match</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>0 rows Ã— 25 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [sample_id, variants, genes, cNomen, pNomen, run, vaf, depth, vcf_mutect, vcf_pindel, variant_caller, vcf_filter, panel, virtual_panel, gnomAD_AF, gnomAD_AF_Pct, cosmic_id, cosmic_haem, run_freq, total_freq, length, var_type, technical_status, retain_discard, status_match]\n",
       "Index: []\n",
       "\n",
       "[0 rows x 25 columns]"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Display all the match rows to show where the equivalence rule is satisfied.\n",
    "\n",
    "match_rows"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "id": "ca2cd6ad-c1f1-407b-9da1-70fd588cdaae",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Write the DataFrame to a CSV file named 'match_rows.csv'.\n",
    "# Use the 'to_csv()' method on the DataFrame 'data' to save it to a CSV file.\n",
    "# Set the 'index' parameter to False to prevent writing row indices to the CSV file.\n",
    "\n",
    "match_rows.to_csv('match_rows.csv', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "ab37df0f-8425-435f-a9dc-4fa73cc5be0e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "all the match_rows saved to 'match_rows.csv'\n"
     ]
    }
   ],
   "source": [
    "# Print the new match rows to the console using the 'print()' function.\n",
    "# Call the 'print()' function with the message: \"all the match_rows saved to 'mismatch_rows.csv'.\n",
    "\n",
    "print(\"all the match_rows saved to 'match_rows.csv'\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "380bab14-607c-428e-8f56-eb858c1affd6",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Filter all the rows where the statuses do not match.\n",
    "# Select rows where `status_match` is `False` and store these rows in a new DataFrame named `mismatch_rows`.\n",
    "\n",
    "mismatch_rows = data[data['status_match'] == False]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "fa7af536-fe21-49d2-ac2b-c1b5e1223390",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Lenght of all the rows where the statuses do not match:\n",
      "100\n"
     ]
    }
   ],
   "source": [
    "# Check the number of all the rows where the statuses do not match in the DataFrame.\n",
    "\n",
    "print(\"Lenght of all the rows where the statuses do not match:\")\n",
    "print(len(mismatch_rows))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "eaeccd9a-954a-48d9-84d6-a34025d17c9d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_id</th>\n",
       "      <th>variants</th>\n",
       "      <th>genes</th>\n",
       "      <th>cNomen</th>\n",
       "      <th>pNomen</th>\n",
       "      <th>run</th>\n",
       "      <th>vaf</th>\n",
       "      <th>depth</th>\n",
       "      <th>vcf_mutect</th>\n",
       "      <th>vcf_pindel</th>\n",
       "      <th>...</th>\n",
       "      <th>gnomAD_AF_Pct</th>\n",
       "      <th>cosmic_id</th>\n",
       "      <th>cosmic_haem</th>\n",
       "      <th>run_freq</th>\n",
       "      <th>total_freq</th>\n",
       "      <th>length</th>\n",
       "      <th>var_type</th>\n",
       "      <th>technical_status</th>\n",
       "      <th>retain_discard</th>\n",
       "      <th>status_match</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Mock001</td>\n",
       "      <td>KRAS_2_160883814_A/G</td>\n",
       "      <td>KRAS</td>\n",
       "      <td>c.1077A&gt;G</td>\n",
       "      <td>p.Ala1367Val</td>\n",
       "      <td>Run1</td>\n",
       "      <td>0.377</td>\n",
       "      <td>354.0</td>\n",
       "      <td>22.0</td>\n",
       "      <td>26.0</td>\n",
       "      <td>...</td>\n",
       "      <td>1.9014</td>\n",
       "      <td>COSM444461</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2</td>\n",
       "      <td>86</td>\n",
       "      <td>0</td>\n",
       "      <td>deletion</td>\n",
       "      <td>Discard</td>\n",
       "      <td>Discard</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Mock002</td>\n",
       "      <td>KRAS_20_183670202_A/G</td>\n",
       "      <td>KRAS</td>\n",
       "      <td>c.2772A&gt;G</td>\n",
       "      <td>p.Ala228Val</td>\n",
       "      <td>Run1</td>\n",
       "      <td>0.727</td>\n",
       "      <td>254.0</td>\n",
       "      <td>29.0</td>\n",
       "      <td>27.0</td>\n",
       "      <td>...</td>\n",
       "      <td>1.1973</td>\n",
       "      <td>COSM989868</td>\n",
       "      <td>6.0</td>\n",
       "      <td>3</td>\n",
       "      <td>84</td>\n",
       "      <td>0</td>\n",
       "      <td>indel</td>\n",
       "      <td>Valid</td>\n",
       "      <td>Retain</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Mock003</td>\n",
       "      <td>KRAS_22_5738866_A/G</td>\n",
       "      <td>KRAS</td>\n",
       "      <td>c.2073A&gt;G</td>\n",
       "      <td>p.Ala218Val</td>\n",
       "      <td>Run3</td>\n",
       "      <td>0.163</td>\n",
       "      <td>291.0</td>\n",
       "      <td>41.0</td>\n",
       "      <td>35.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.3120</td>\n",
       "      <td>COSM924050</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0</td>\n",
       "      <td>80</td>\n",
       "      <td>0</td>\n",
       "      <td>deletion</td>\n",
       "      <td>Discard</td>\n",
       "      <td>Discard</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Mock004</td>\n",
       "      <td>BRCA2_18_377694477_A/G</td>\n",
       "      <td>BRCA2</td>\n",
       "      <td>c.2030A&gt;G</td>\n",
       "      <td>p.Ala947Val</td>\n",
       "      <td>Run3</td>\n",
       "      <td>0.067</td>\n",
       "      <td>619.0</td>\n",
       "      <td>39.0</td>\n",
       "      <td>50.0</td>\n",
       "      <td>...</td>\n",
       "      <td>1.7324</td>\n",
       "      <td>COSM606687</td>\n",
       "      <td>2.0</td>\n",
       "      <td>3</td>\n",
       "      <td>68</td>\n",
       "      <td>0</td>\n",
       "      <td>substitution</td>\n",
       "      <td>Valid</td>\n",
       "      <td>Discard</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Mock005</td>\n",
       "      <td>ARID1A_15_547373809_A/G</td>\n",
       "      <td>ARID1A</td>\n",
       "      <td>c.382A&gt;G</td>\n",
       "      <td>p.Ala782Val</td>\n",
       "      <td>Run2</td>\n",
       "      <td>0.599</td>\n",
       "      <td>718.0</td>\n",
       "      <td>33.0</td>\n",
       "      <td>14.0</td>\n",
       "      <td>...</td>\n",
       "      <td>1.4161</td>\n",
       "      <td>COSM938076</td>\n",
       "      <td>9.0</td>\n",
       "      <td>0</td>\n",
       "      <td>11</td>\n",
       "      <td>0</td>\n",
       "      <td>indel</td>\n",
       "      <td>Discard</td>\n",
       "      <td>Retain</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>95</th>\n",
       "      <td>Mock096</td>\n",
       "      <td>ARID1A_15_103119975_A/G</td>\n",
       "      <td>ARID1A</td>\n",
       "      <td>c.2147A&gt;G</td>\n",
       "      <td>p.Ala148Val</td>\n",
       "      <td>Run3</td>\n",
       "      <td>0.101</td>\n",
       "      <td>600.0</td>\n",
       "      <td>19.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>1.7944</td>\n",
       "      <td>COSM719669</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3</td>\n",
       "      <td>64</td>\n",
       "      <td>0</td>\n",
       "      <td>indel</td>\n",
       "      <td>Valid</td>\n",
       "      <td>Retain</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>96</th>\n",
       "      <td>Mock097</td>\n",
       "      <td>KMT2C_21_183025592_A/G</td>\n",
       "      <td>KMT2C</td>\n",
       "      <td>c.1946A&gt;G</td>\n",
       "      <td>p.Ala238Val</td>\n",
       "      <td>Run3</td>\n",
       "      <td>0.892</td>\n",
       "      <td>477.0</td>\n",
       "      <td>21.0</td>\n",
       "      <td>36.0</td>\n",
       "      <td>...</td>\n",
       "      <td>1.2662</td>\n",
       "      <td>COSM868871</td>\n",
       "      <td>6.0</td>\n",
       "      <td>3</td>\n",
       "      <td>18</td>\n",
       "      <td>0</td>\n",
       "      <td>substitution</td>\n",
       "      <td>Discard</td>\n",
       "      <td>Retain</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>97</th>\n",
       "      <td>Mock098</td>\n",
       "      <td>DNMT3A_9_398039145_A/G</td>\n",
       "      <td>DNMT3A</td>\n",
       "      <td>c.185A&gt;G</td>\n",
       "      <td>p.Ala170Val</td>\n",
       "      <td>Run3</td>\n",
       "      <td>0.342</td>\n",
       "      <td>257.0</td>\n",
       "      <td>34.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.6984</td>\n",
       "      <td>COSM677026</td>\n",
       "      <td>3.0</td>\n",
       "      <td>0</td>\n",
       "      <td>86</td>\n",
       "      <td>0</td>\n",
       "      <td>deletion</td>\n",
       "      <td>Valid</td>\n",
       "      <td>Discard</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>98</th>\n",
       "      <td>Mock099</td>\n",
       "      <td>ANA10_19_419898203_A/G</td>\n",
       "      <td>ANA10</td>\n",
       "      <td>c.2545A&gt;G</td>\n",
       "      <td>p.Ala1129Val</td>\n",
       "      <td>Run2</td>\n",
       "      <td>0.721</td>\n",
       "      <td>159.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>37.0</td>\n",
       "      <td>...</td>\n",
       "      <td>1.7942</td>\n",
       "      <td>COSM541803</td>\n",
       "      <td>5.0</td>\n",
       "      <td>3</td>\n",
       "      <td>88</td>\n",
       "      <td>0</td>\n",
       "      <td>indel</td>\n",
       "      <td>Valid</td>\n",
       "      <td>Retain</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>99</th>\n",
       "      <td>Mock100</td>\n",
       "      <td>DNMT3A_11_88989766_A/G</td>\n",
       "      <td>DNMT3A</td>\n",
       "      <td>c.1752A&gt;G</td>\n",
       "      <td>p.Ala550Val</td>\n",
       "      <td>Run2</td>\n",
       "      <td>0.879</td>\n",
       "      <td>482.0</td>\n",
       "      <td>33.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>...</td>\n",
       "      <td>1.5598</td>\n",
       "      <td>COSM469584</td>\n",
       "      <td>5.0</td>\n",
       "      <td>2</td>\n",
       "      <td>65</td>\n",
       "      <td>0</td>\n",
       "      <td>indel</td>\n",
       "      <td>Discard</td>\n",
       "      <td>Retain</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>100 rows Ã— 25 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   sample_id                 variants   genes     cNomen        pNomen   run  \\\n",
       "0    Mock001     KRAS_2_160883814_A/G    KRAS  c.1077A>G  p.Ala1367Val  Run1   \n",
       "1    Mock002    KRAS_20_183670202_A/G    KRAS  c.2772A>G   p.Ala228Val  Run1   \n",
       "2    Mock003      KRAS_22_5738866_A/G    KRAS  c.2073A>G   p.Ala218Val  Run3   \n",
       "3    Mock004   BRCA2_18_377694477_A/G   BRCA2  c.2030A>G   p.Ala947Val  Run3   \n",
       "4    Mock005  ARID1A_15_547373809_A/G  ARID1A   c.382A>G   p.Ala782Val  Run2   \n",
       "..       ...                      ...     ...        ...           ...   ...   \n",
       "95   Mock096  ARID1A_15_103119975_A/G  ARID1A  c.2147A>G   p.Ala148Val  Run3   \n",
       "96   Mock097   KMT2C_21_183025592_A/G   KMT2C  c.1946A>G   p.Ala238Val  Run3   \n",
       "97   Mock098   DNMT3A_9_398039145_A/G  DNMT3A   c.185A>G   p.Ala170Val  Run3   \n",
       "98   Mock099   ANA10_19_419898203_A/G   ANA10  c.2545A>G  p.Ala1129Val  Run2   \n",
       "99   Mock100   DNMT3A_11_88989766_A/G  DNMT3A  c.1752A>G   p.Ala550Val  Run2   \n",
       "\n",
       "      vaf  depth  vcf_mutect  vcf_pindel  ... gnomAD_AF_Pct   cosmic_id  \\\n",
       "0   0.377  354.0        22.0        26.0  ...        1.9014  COSM444461   \n",
       "1   0.727  254.0        29.0        27.0  ...        1.1973  COSM989868   \n",
       "2   0.163  291.0        41.0        35.0  ...        0.3120  COSM924050   \n",
       "3   0.067  619.0        39.0        50.0  ...        1.7324  COSM606687   \n",
       "4   0.599  718.0        33.0        14.0  ...        1.4161  COSM938076   \n",
       "..    ...    ...         ...         ...  ...           ...         ...   \n",
       "95  0.101  600.0        19.0         1.0  ...        1.7944  COSM719669   \n",
       "96  0.892  477.0        21.0        36.0  ...        1.2662  COSM868871   \n",
       "97  0.342  257.0        34.0         1.0  ...        0.6984  COSM677026   \n",
       "98  0.721  159.0         1.0        37.0  ...        1.7942  COSM541803   \n",
       "99  0.879  482.0        33.0         4.0  ...        1.5598  COSM469584   \n",
       "\n",
       "   cosmic_haem run_freq  total_freq  length      var_type technical_status  \\\n",
       "0          0.0        2          86       0      deletion          Discard   \n",
       "1          6.0        3          84       0         indel            Valid   \n",
       "2          2.0        0          80       0      deletion          Discard   \n",
       "3          2.0        3          68       0  substitution            Valid   \n",
       "4          9.0        0          11       0         indel          Discard   \n",
       "..         ...      ...         ...     ...           ...              ...   \n",
       "95         0.0        3          64       0         indel            Valid   \n",
       "96         6.0        3          18       0  substitution          Discard   \n",
       "97         3.0        0          86       0      deletion            Valid   \n",
       "98         5.0        3          88       0         indel            Valid   \n",
       "99         5.0        2          65       0         indel          Discard   \n",
       "\n",
       "    retain_discard  status_match  \n",
       "0          Discard         False  \n",
       "1           Retain         False  \n",
       "2          Discard         False  \n",
       "3          Discard         False  \n",
       "4           Retain         False  \n",
       "..             ...           ...  \n",
       "95          Retain         False  \n",
       "96          Retain         False  \n",
       "97         Discard         False  \n",
       "98          Retain         False  \n",
       "99          Retain         False  \n",
       "\n",
       "[100 rows x 25 columns]"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Display all the mismatch rows to show where the equivalence rule is violated.\n",
    "\n",
    "mismatch_rows"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "id": "7268b65e-03bb-45c9-b720-f3d60aef14dd",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Write the DataFrame to a CSV file named 'mismatch_rows.csv'.\n",
    "# Use the 'to_csv()' method on the DataFrame 'data' to save it to a CSV file.\n",
    "# Set the 'index' parameter to False to prevent writing row indices to the CSV file.\n",
    "\n",
    "mismatch_rows.to_csv('mismatch_rows.csv', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "530de6f9-e106-48dd-a867-b4e684ee3da8",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "all the mismatch_rows saved to 'mismatch_rows.csv'\n"
     ]
    }
   ],
   "source": [
    "# Print the new mismatch rows to the console using the 'print()' function.\n",
    "# Call the 'print()' function with the message: \"all the mismatch_rows saved to 'mismatch_rows.csv'.\n",
    "\n",
    "print(\"all the mismatch_rows saved to 'mismatch_rows.csv'\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [conda env:base] *",
   "language": "python",
   "name": "conda-base-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
